File: leetcode/1120. Maximum Average Subtree/1120.cpp
// https://helloacm.com/the-maximum-average-subtree-of-a-binary-tree/
// https://leetcode.com/problems/maximum-average-subtree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    double maximumAverageSubtree(TreeNode* root) {
        int cnt = 0;
        getSum(root, cnt);
        return maxAvg;
    }
    
private:
    int getSum(TreeNode* root, int &cnt) {
        if (root == nullptr) return 0;
        int left_cnt = 0;
        int right_cnt = 0;
        int left = getSum(root->left, left_cnt);
        int right = getSum(root->right, right_cnt);
        int sum = left + right + root->val;
        cnt = left_cnt + right_cnt + 1;
        maxAvg = max(maxAvg, (double)sum / cnt);
        return sum;
    }
    double maxAvg = 0.0;
};

File: leetcode/1120. Maximum Average Subtree/1120.py
# https://helloacm.com/teaching-kids-programming-recursive-depth-first-search-algorithm-to-compute-the-max-average-of-a-binary-subtree/
# https://leetcode.com/problems/maximum-average-subtree/
# MEDIUM, DFS

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maximumAverageSubtree(self, root: Optional[TreeNode]) -> float:
        ans = 0        
        def dfs(root):
            if not root:
                return (0, 0)
            nonlocal ans
            left, cnt1 = dfs(root.left)
            right, cnt2 = dfs(root.right)
            c = cnt1 + cnt2 + 1
            s = left + right + root.val
            ans = max(ans, s / c)
            return (s, c)
        dfs(root)
        return ans
